dcap/types/src/tcb.rs

```diff
@@ -0,0 +1,362 @@
+    ObjectIdentifier::new_unwrap("1.2.840.113741.1.13.1.2.1"),
+    ObjectIdentifier::new_unwrap("1.2.840.113741.1.13.1.2.2"),
+    ObjectIdentifier::new_unwrap("1.2.840.113741.1.13.1.2.3"),
+    ObjectIdentifier::new_unwrap("1.2.840.113741.1.13.1.2.4"),
+    ObjectIdentifier::new_unwrap("1.2.840.113741.1.13.1.2.5"),
+    ObjectIdentifier::new_unwrap("1.2.840.113741.1.13.1.2.6"),
+    ObjectIdentifier::new_unwrap("1.2.840.113741.1.13.1.2.7"),
+    ObjectIdentifier::new_unwrap("1.2.840.113741.1.13.1.2.8"),
+    ObjectIdentifier::new_unwrap("1.2.840.113741.1.13.1.2.9"),
+    ObjectIdentifier::new_unwrap("1.2.840.113741.1.13.1.2.10"),
+    ObjectIdentifier::new_unwrap("1.2.840.113741.1.13.1.2.11"),
+    ObjectIdentifier::new_unwrap("1.2.840.113741.1.13.1.2.12"),
+    ObjectIdentifier::new_unwrap("1.2.840.113741.1.13.1.2.13"),
+    ObjectIdentifier::new_unwrap("1.2.840.113741.1.13.1.2.14"),
+    ObjectIdentifier::new_unwrap("1.2.840.113741.1.13.1.2.15"),
+    ObjectIdentifier::new_unwrap("1.2.840.113741.1.13.1.2.16"),
+];
+const PCE_SVN_OID: ObjectIdentifier = ObjectIdentifier::new_unwrap("1.2.840.113741.1.13.1.2.17");
+const FMSPC_OID: ObjectIdentifier = ObjectIdentifier::new_unwrap("1.2.840.113741.1.13.1.4");
+
+#[derive(Debug, PartialEq)]
+pub enum Error {
+    MissingSgxExtension(String),
+    Der(x509_cert::der::Error),
+    FmspcSize(usize),
+}
+
+impl From<x509_cert::der::Error> for Error {
+    fn from(err: x509_cert::der::Error) -> Self {
+        Error::Der(err)
+    }
+}
+
+// The SGX extensions aren't really documented. They aren't RFC 5280 extensions
+// which are `OID` and `OCTET`. They're an `OID` and an `Any`, which is what the
+// [`AttributeTypeAndValue`] is.
+type SgxExtensions = Vec<AttributeTypeAndValue>;
+
+/// The TCB info provided by the PCK(Provisioning Certification Key) leaf
+/// certificate
+#[derive(Debug, PartialEq)]
+pub struct PckTcbInfo {
+    svns: [u32; 16],
+    pce_svn: u32,
+    fmspc: [u8; FMSPC_SIZE],
+}
+
+impl TryFrom<&Certificate> for PckTcbInfo {
+    type Error = Error;
+
+    fn try_from(cert: &Certificate) -> Result<Self, Self::Error> {
+        let sgx_extensions = sgx_extensions(cert)?;
+
+        let fmspc = fmspc(&sgx_extensions)?;
+
+        let (pce_svn, svns) = tcb_svns(&sgx_extensions)?;
+
+        Ok(PckTcbInfo {
+            svns,
+            pce_svn,
+            fmspc,
+        })
+    }
+}
+
+/// Get the [`SgxExtensions`] from the `cert`.
+///
+/// # Errors
+/// * `Error::MissingSgxExtension` if the `cert` does not have the SGX extension.
+/// * `Error::DerDecoding` if the contained DER is invalid.
+fn sgx_extensions(cert: &Certificate) -> Result<SgxExtensions, Error> {
+    let extensions = &cert.tbs_certificate.extensions;
+    let extension = match extensions
+        .iter()
+        .flatten()
+        .find(|extension| extension.extn_id == SGX_OID)
+    {
+        Some(extension) => extension,
+        None => return Err(Error::MissingSgxExtension(SGX_OID.to_string())),
+    };
```

Comment:
Nit: I believe this can be shortened to `.ok_or_else(|| Error::MissingSgxExtension(SGX_OID.to_string()))?`

Reply:
thanks, like the shorter version